### ISSUE TITLE
Document style of "support" comments

### ISSUE THREAD

**Issue by markelog** [jquery/contribute.jquery.org#95]
Every time we have a rewrite or drop support for some browser we engage (at least in Core) in discussion of how to properly write those support comments:

> Support: IE8<9
> IE8+
> IE11<
> ...

We need clear understanding of how to write those comments, so we can leave those discussion behind and focus on code instead of flaming the pull/issue thread with pointless posts.

**Comment by scottgonzalez** [comment]
Do we have consensus yet?

**Comment by gibson042** [comment]
> Do we have consensus yet?

Comprehensive strawman to check ([RFC 5234](https://tools.ietf.org/html/rfc5234) / [RFC 7405](https://tools.ietf.org/html/rfc7405) ABNF syntax):

```
support-comments     = 1*support-comment
                           ; code addressing multiple bugs SHOULD be preceded by
                           ; a support comment for each

support-comment      = <affects> LF <description>
                           ; to be followed immediately by code (no intervening blank line)

affects              = indent
                       %s"// Support: " package-combination *( ", " package-combination )
                       [ affects-more ]
                           ; case-sensitive "Support" followed by affected packages

affects-more         = "," LF indent
                       "// " *WSP package-combination *( ", " package-combination )
                       [ affects-more ]
                           ; trailing comma on the previous line

package-combination  = package-range *( %s" with " package-range )
                           ; "with"-separated list of version-ranged packages

package-range        =  package
                           ; unversioned "package" (e.g. "T-Mobile")

package-range        =  package SP version [ " only" ]
                           ; single version

package-range        =/ package SP version " - " version [ " only" ]
                           ; fixed range

package-range        =/ package SP "<=" version " - " version [ " only" ]
                           ; fixed range with unknown start

package-range        =/ package SP ( "<=" version / "<" version " only" )
                           ; open range

package              =  1*ALPHA
                           ; most names are simple

package              =/ VCHAR / VCHAR *( SP / VCHAR ) VCHAR
                           ; ...but weirdness is allowed

version              =  1*DIGIT 0*2( "." 1*DIGIT )
                           ; most versions are simple

version              =/ VCHAR / VCHAR *( SP / VCHAR ) VCHAR
                           ; ...but weirdness is allowed

description          = indent "// " VCHAR *( WSP / VCHAR ) LF *description-more
                           ; arbitrary string of space/tab/printable ASCII explaining
                           ;   either the problem or the solution
                           ; first VCHAR MUST NOT be a lowercase letter
                           ; each line not to exceed 100 visual characters in total length,
                           ;   per http://contribute.jquery.org/style-guide/js/#spacing

description-more     = indent "// " 1*( WSP / VCHAR ) LF

indent               (MUST match context)
```

Examples:
```js
// Support: IE < 9 only
// Tolerate nodeList.length returning elements with id "length"
// Support: Safari <= 6
// Tolerate nodeList[ number ] returning elements with id equal to the number
// (jQuery #14142)
for ( ; i !== len && (elem = elems[i]) != null; i++ ) {
```

```js
// Support: Chrome < 29 only, Android <= 4.2, Safari <= 7.0, iOS <= 7.0,
//   PhantomJS <= 1.9.7
// `~=` doesn't recognize non-space non-tab whitespace
if ( !div.querySelectorAll("[id~=d]").length ) {
```

```js
// Support: Android <=2.3 - 4.1 only, Node.js with jsdom <= 1.5.0,
//   Opera < 15 only with jQuery 1.4 - 1.7 only
// Long description
// (use your imagination)
```

Detailed explanation of `package-range`:

Introduced\ Fixed | `n` (immediately follows `m`) | not fixed as of `n` (latest checked)
------------------ |--------------- | -------------------------------------
in or before `k` (implied) |  `package <n only`   | `package <=n`
in or before `k` (explicit) |  `package <=k - m only`   | `package <=k - n`
`k`                | `package k - m only`  | `package k - n`
`m`                | `package m only` | `package m - n` (same as above)
`n`                | _(meaningless)_ | `package n`

**Comment by dmethvin** [comment]
OH NO HE DID-UNT! :astonished:

Most of it seems pretty straightforward, but I'm not clear on the use of `only` for open-ended bugs that haven't been resolved in a current browser.

Now we just need JSCS to implement the parser and check the style so we always get it right. :smiling_imp:

**Comment by scottgonzalez** [comment]
Open-ended bugs that haven't been resolved in a current browser don't get `only`. Note that the rightmost column never has `only`.

**Comment by dmethvin** [comment]
Agreed, the table makes it clear. I was trying to follow the grammar for a package-range.

**Comment by markelog** [comment]
> Now we just need JSCS to implement the parser 

That rule seems way to specific to include it in the jquery preset, would it be fine if i do it as jscs plugin?

**Comment by dmethvin** [comment]
Sure, I agree it's not a general purpose thing.

**Comment by gibson042** [comment]
Looking at it again, we should probably drop the snowflake `<n only` in favor of `<=m only`.

**Comment by gibson042** [comment]
> Looking at it again, we should probably drop the snowflake `<n only` in favor of `<=m only`.

Done.

**Comment by dmethvin** [comment]
Discussed and agreed on in the core meeting. For the prose, we'll just summarize what's in the discussion above and add the table. The grammar can go into a separate page.

**Comment by scottgonzalez** [comment]
Did anyone volunteer to send a PR?

**Comment by markelog** [comment]
How about @gibson042 will send PR for contribute and will create a plugin?

**Comment by markelog** [comment]
But i'd say it's not a priority, 3.0 beta stuff first

**Comment by dmethvin** [comment]
As priorities go I'd rather focus on getting core 3.0 out and we can circle back to formally documenting this. At least we have this issue now.

**Comment by gibson042** [comment]
> I'd rather focus on getting core 3.0 out and we can circle back to formally documenting this

Agreed. It may turn out that I'm the one to do it, but certainly not before core 3.0.

**Comment by markelog** [comment]
> and will create a plugin?

Sorry missed "i" here, should be:

> and i will create a plugin?

**Comment by jzaefferer** [comment]
@markelog I'd be happy to test that plugin on jQuery UI and QUnit, let me know when you have something. Might be good to implement that plugin before doing the contribute PR, could highlight some issues in the spec.

**Comment by markelog** [comment]
@jzaefferer oh, missed your comment, will try tackle this one.

**Comment by mzgol** [comment]
How do we want to write about Microsoft Edge? Since IE has no further versions planned, I guess we should not write:
```
IE <= 6-11
```
but:
```
IE <= 6-11 only, Edge ?
```
The question is what to write instead of `?` considering that Edge has no versions unlike Firefox/Chrome.

Build numbers? That could work if they are increasing integers. If they're not, maybe dates?

(I wonder if BrowserStack will make more than 1 Edge build available).

**Comment by markelog** [comment]
I guess build numbers, date might not be too descriptive and seriously, no versions? Could you give me a link?

**Comment by mzgol** [comment]
> Could you give me a link?

https://twitter.com/jacobrossi/status/607060127837908992

**Comment by mzgol** [comment]
We need to start documenting workarounds for IE if they're needed in Edge as well so I'd like to at least settle this. It seems Edge has increasing build numbers and they're available in the user agent in the form: `Edge/12.BUILD_NUMBER`. We should probably just use those build numbers since the `12.` part ["is mostly meaningless"](https://github.com/jquery/jquery/commit/8e111df641cca3e1b75b31a1971bfc89014b4ded#commitcomment-11850878). It will look weird to write:
```
// Support: IE 10-11 only, Edge 9600-10159+
```
but we'll get used to it. ;)

Another thing: should we write `IE 10-11 only, Edge 9600-10159+` or `IE 10-11, Edge 9600-10159+`? It fits both since there will be no IE12.

**Comment by scottgonzalez** [comment]
I'd recommend writing the full version, including the 12.

**Comment by mzgol** [comment]
This would be even more verbose: `Edge 12.9600-12.10159+`. The build numbers are exactly Windows 10 build numbers so they're not magical. If Edge changes the `12` at all in the future, we can always introduce another number but the lone fact that `12.` appears in the user agent doesn't need to mean anything.

**Comment by dmethvin** [comment]
Right now the Edge and Win10 numbers are synced but I wonder if it will stay that way after final release. I'd prefer something short. We should pull in @jacobrossi here just to be sure we're doing it in a way that will make sense once they start to change numbers.

**Comment by jacobrossi** [comment]
Yep, what you're doing makes sense and should be fine. The numbers are increasing and in sync with the OS and we plan to continue that.

Note there's still a chance we could fix a thing or two through Windows Update and this build number wouldn't change (e.g. like the IE11 case mentioned above).  We can't include patch/revision numbers that change in the monthly WU updates for security reasons.

**Comment by mzgol** [comment]
@jacobrossi 

Thanks for the reply!

> We can't include patch/revision numbers that change in the monthly WU updates for security reasons.

I thought each WU constitutes a new Windows build. How often will the build number change? I'm interested only very approximately, like, each month, each 6 months, each year, each 3 years etc.?

**Comment by mzgol** [comment]
> This would be even more verbose: `Edge 12.9600-12.10159+`.

Yet another reason to write it like that - these numbers are not unambiguous. For example, the latest stable build, 10240 has Edge with `Edge/12.10240` in the user agent string but `Microsoft Edge 20.10240.16384.0` on the Settings pane. The only unambiguous part is the build number; 10240 in both cases.

BTW, @jacobrossi, why do those numbers differ so greatly? It's going to get confusing when some tools choose the UA version and others the only from the Settings pane.

**Comment by mzgol** [comment]
In light of new info at https://github.com/jquery/testswarm/issues/308#issuecomment-142134534 we should use `Edge 12 only`, `Edge 12-13+` etc.

**Comment by gibson042** [comment]
Works for me. It reads a little strangely when the "plus" version doesn't exist, but remains comprehensible.

**Comment by jbedard** [comment]
I agree `Firefox <=35-42` looks like the issue no longer exists in `43+` and would suggest deleting something once `42` is no longer supported. Your suggestion SGTM 👍
